### PR TITLE
adding auto-pilot support

### PR DESCRIPTION
**Pull request by cmcga1125**
## autopilot support

hey friends,
Here's my PR addressing issue #835 enabling [GKE Autopilot Support](https://cloud.google.com/blog/products/containers-kubernetes/introducing-gke-autopilot). right now it is [blocked by an upstream PR to the provider](https://github.com/hashicorp/terraform-provider-google/pull/8632), but this should work and i can test a cluster once the upstream PR is merged.  Its a pretty straight forward change, however, be aware that enabling the feature kills sections of needed code like node pools.  Additionally it has some more [cluster limitations](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#unsupported_cluster_features)

### CODE REVIEW

**Comment by cmcga1125** [comment]
also, your make commands are 🔥

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Attempting to download /workspace/test/bundle.hcl bundle.
/tmp/bundler /workspace
Fetching Terraform 0.13.6 core package...
Local plugin directory ".plugins" found; scanning for provider binaries.
No ".plugins" directory found, skipping local provider discovery.
Creating terraform_0.13.6-bundle2021031615_linux_amd64.zip ...
All done!
Archive:  terraform_0.13.6-bundle2021031615_linux_amd64.zip
/workspace
Running terraform fmt
Running terraform validate
terraform_validate . 
Error: 
Unsupported argument
  on cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/deploy_service 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/disable_client_cert 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/node_pool 
Error: 
Unsupported argument
  on ../../modules/beta-public-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/node_pool_update_variant 
Error: 
Unsupported argument
  on ../../modules/private-cluster-update-variant/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/node_pool_update_variant_beta 
Error: 
Unsupported argument
  on ../../modules/beta-private-cluster-update-variant/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/node_pool_update_variant_public_beta 
Error: 
Unsupported argument
  on ../../modules/beta-public-cluster-update-variant/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/private_zonal_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on main.tf line 42, in module "gcp-network":
  42:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/regional_private_node_pool_oauth_scopes 
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/safer_cluster 
Warning: 
Interpolation-only expressions are deprecated
  on network.tf line 37, in module "gcp-network":
  37:     
"${local.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../modules/beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/safer_cluster_iap_bastion 
Warning: 
Interpolation-only expressions are deprecated
  on network.tf line 36, in module "vpc":
  36:     
"${var.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../modules/beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/shared_vpc 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional_beta 
Error: 
Unsupported argument
  on ../../modules/beta-public-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional_private 
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional_private_beta 
Error: 
Unsupported argument
  on ../../modules/beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional_with_kubeconfig 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_regional_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on main.tf line 45, in module "gcp-network":
  45:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_zonal_private 
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_zonal_with_acm 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_zonal_with_asm 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_zonal_with_hub 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/simple_zonal_with_hub_kubeconfig 
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/stub_domains_private 
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/stub_domains_upstream_nameservers 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/upstream_nameservers 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/workload_identity 
Error: 
Unsupported argument
  on ../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./examples/workload_metadata_config 
Error: 
Unsupported argument
  on ../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/acm 
Success!
 The configuration is valid.
terraform_validate ./modules/asm 
Success!
 The configuration is valid.
terraform_validate ./modules/auth 
Success!
 The configuration is valid.
terraform_validate ./modules/beta-private-cluster 
Error: 
Unsupported argument
  on cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/beta-private-cluster-update-variant 
Error: 
Unsupported argument
  on cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/beta-public-cluster 
Error: 
Unsupported argument
  on cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/beta-public-cluster-update-variant 
Error: 
Unsupported argument
  on cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/binary-authorization 
Success!
 The configuration is valid.
terraform_validate ./modules/config-sync 
Success!
 The configuration is valid.
terraform_validate ./modules/hub 
Success!
 The configuration is valid.
terraform_validate ./modules/k8s-operator-crd-support 
Success!
 The configuration is valid.
terraform_validate ./modules/private-cluster 
Error: 
Unsupported argument
  on cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/private-cluster-update-variant 
Error: 
Unsupported argument
  on cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/safer-cluster 
Error: 
Unsupported argument
  on ../beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/safer-cluster-update-variant 
Error: 
Unsupported argument
  on ../beta-private-cluster-update-variant/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./modules/services 
Success!
 The configuration is valid.
terraform_validate ./modules/workload-identity 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/beta_cluster 
Error: 
Unsupported argument
  on ../../../modules/beta-public-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/deploy_service 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/disable_client_cert 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/node_pool 
Error: 
Unsupported argument
  on ../../../modules/beta-public-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/node_pool_update_variant 
Error: 
Unsupported argument
  on ../../../modules/private-cluster-update-variant/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/private_zonal_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/private_zonal_with_networking/main.tf line 42, in module "gcp-network":
  42:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/safer_cluster 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/safer_cluster/network.tf line 37, in module "gcp-network":
  37:     
"${local.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../../modules/beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/safer_cluster_iap_bastion 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/safer_cluster_iap_bastion/network.tf line 36, in module "vpc":
  36:     
"${var.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../../modules/beta-private-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/sandbox_enabled 
Error: 
Unsupported argument
  on ../../../modules/beta-public-cluster/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_regional 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_regional_private 
Error: 
Unsupported argument
  on ../../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_regional_with_kubeconfig 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_regional_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/simple_regional_with_networking/main.tf line 45, in module "gcp-network":
  45:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_zonal 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_zonal_private 
Error: 
Unsupported argument
  on ../../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/simple_zonal_with_asm 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/stub_domains 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/stub_domains_private 
Error: 
Unsupported argument
  on ../../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/stub_domains_upstream_nameservers 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/upstream_nameservers 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/workload_identity 
Error: 
Unsupported argument
  on ../../../cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/fixtures/workload_metadata_config 
Error: 
Unsupported argument
  on ../../../modules/private-cluster/cluster.tf line 78, in resource "google_container_cluster" "primary":
  78:   
enable_autopilot
            = var.enable_autopilot
An argument named "enable_autopilot" is not expected here.
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

**Comment by cmcga1125** [comment]
tried building an autopilot private cluster and got this - i'll wait until the upstream is merged and try again - it might be good to have an example as it has some caveats?

```
Error: Unsupported argument

  on .terraform/modules/gke/modules/beta-private-cluster-update-variant/cluster.tf line 96, in resource "google_container_cluster" "primary":
  96:   enable_autopilot            = var.enable_autopilot

An argument named "enable_autopilot" is not expected here.
```

**Comment by morgante** [comment]
I'm wondering if autopilot clusters should actually be a distinct submodule, since they seem fairly different. Ex. node pool configuration isn't applicable.

@bharathkkb Thoughts?

**Comment by bharathkkb** [comment]
@morgante I think autopilot should be a submodule, we can should be able to adapt the existing tmpl to support this.

**Comment by cmcga1125** [comment]
yep - was torn on if it should be separate, but the provider did not break from the standard `google_container_cluster` resource and from what i can tell you can just omit the node pool.  I do get the whole make it separate argument.  I can give it ago - looking for the configuration that splits out new modules 🙃

**Comment by morgante** [comment]
The modules config is here: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/autogen_modules.json

**Comment by cmcga1125** [comment]
i tried to mark the PR as wip or something but don't seem to have permissions 😃 

Also, i think its in a decent place to test once the provider drops. i'll keep an eye out -- thanks for the opportunity, i've been wanting to commit to some public projects :)

**Comment by morgante** [comment]
@cmcga1125 Thanks for working on it! I've gone ahead and converted your PR to a draft for now.

**Comment by cmcga1125** [comment]
as an update - the provider is hoping to have this out march 29 2021

**Comment by cmcga1125** [comment]
Successfully created a private cluster :) 
I'll test a public cluster in a bit.

**Comment by cmcga1125** [comment]
I've successfully built both cluster types with this module - but the `kubernetes-engine-lint-trigger` is failing.  i can't see the test output to correct.  Can someone provide?

**Comment by morgante** [comment]
@cmcga1125 The lint bot updates [this comment](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/850#issuecomment-800359298) with the results. You can also run tests locally: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/CONTRIBUTING.md

**Comment by bharathkkb** [comment]
@cmcga1125 looks like the lint test is passing https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/850/checks?check_run_id=2328026190. Comment is not being updated due to a timeout issue.

**Comment by cmcga1125** [comment]
any idea when you all could look at this and/or consider merging? thanks!!

**Comment by morgante** [comment]
@cmcga1125 Do you think you could add an example/test? It will make merging easier.

**Comment by cmcga1125** [comment]
yep, give me a couple of days

File: modules/beta-autopilot-private-cluster/cluster.tf [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        content {
+          cidr_block   = lookup(cidr_blocks.value, "cidr_block", "")
+          display_name = lookup(cidr_blocks.value, "display_name", "")
+        }
+      }
+    }
+  }
+
+  master_auth {
+    username = var.basic_auth_username
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/cluster.tf:76]: I dont think it is possible to use basic auth and will anyway be removed 1.19+, we could drop this.
@morgante any thoughts?

File: modules/beta-autopilot-private-cluster/cluster.tf [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      }
+    }
+
+    kalm_config {
+      enabled = var.kalm_config
+    }
+
+    config_connector_config {
+      enabled = var.config_connector
+    }
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/cluster.tf:124]: We may need to validate these. Looks like istio, cloudrun_config,config_connector_config,kalm are not supported.
`dns_cache_config` has to be true.
```
googleapi: Error 400: Addons {"istio"} are not supported for Autopilot clusters., badRequest
```

File: modules/beta-autopilot-private-cluster/cluster.tf [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        exclusion_name = maintenance_exclusion.value.name
+        start_time     = maintenance_exclusion.value.start_time
+        end_time       = maintenance_exclusion.value.end_time
+      }
+    }
+
+  }
+
+  lifecycle {
+    ignore_changes = [node_pool, initial_node_count]
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/cluster.tf:163]: do we need to ignore changes? I believe this was workaround for permadiff with initial nodepools in standard mode

File: modules/beta-autopilot-private-cluster/cluster.tf [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    for_each = var.database_encryption
+
+    content {
+      key_name = database_encryption.value.key_name
+      state    = database_encryption.value.state
+    }
+  }
+}
+
+/******************************************
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/cluster.tf:217]: nit: lets add conditional before this so this is not autogenerated

File: modules/beta-autopilot-private-cluster/cluster.tf [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+      dynamic "master_global_access_config" {
+        for_each = var.master_global_access_enabled ? [var.master_global_access_enabled] : []
+        content {
+          enabled = master_global_access_config.value
+        }
+      }
+    }
+  }
+
+  dynamic "database_encryption" {
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/cluster.tf:207]: dont this this is supported for autopilot

File: modules/beta-autopilot-private-cluster/variables.tf [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+  type        = string
+  description = "The region to host the cluster in (optional if zonal cluster / required if regional)"
+  default     = null
+}
+
+variable "zones" {
+  type        = list(string)
+  description = "The zones to host the cluster in (optional if regional cluster / required if zonal)"
+  default     = []
+}
```
Comment (bharathkkb) [inline modules/beta-autopilot-private-cluster/variables.tf:51]: Since ap is only regional, we can just have region?

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR, I took a first pass at the generated module.
Could we enable one of the added examples in the [tests](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/CONTRIBUTING.md#integration-testing) so we can confirm it spins up in autopilot mode.

File: modules/beta-autopilot-private-cluster/cluster.tf (hunk 1 above)
Comment (cmcga1125) [reply modules/beta-autopilot-private-cluster/cluster.tf:76]: separate PR/issue?

File: modules/beta-autopilot-private-cluster/cluster.tf (hunk 2 above)
Comment (cmcga1125) [reply modules/beta-autopilot-private-cluster/cluster.tf:124]: just flipped this chunk off as dns_cache_config by default is set to true

File: modules/beta-autopilot-private-cluster/variables.tf (hunk 6 above)
Comment (cmcga1125) [reply modules/beta-autopilot-private-cluster/variables.tf:51]: working this out would either cause a split in logic {if statement based on autopilot_cluster} or a rewrite of the  [locals in the main.tf](https://github.com/cmcga1125/terraform-google-kubernetes-engine/blob/cmcga1125/835/modules/beta-autopilot-private-cluster/main.tf#L36) file.

another more simple solution would be to set the variable regional to `true` then remove it from the examples/documentation for this cluster?  What do ya'll think?

**Comment by morgante** [comment]
@cmcga1125 Are you still planning to work on this? We'd love to see autopilot support landed. :)

**Comment by cmcga1125** [comment]
hey! sorry i got very busy and am struggling to get the tests to run correctly.  any guidance would be helpful :)

**Comment by morgante** [comment]
What issues are you having with the tests? This [codelab](https://codelabs.developers.google.com/codelabs/cft-onboarding/#0) might help to understand how they work.

**Comment by cmcga1125** [comment]
i'll run through the codelab tonight and see if that helps.

**Comment by eana** [comment]
@cmcga1125 Did you have any luck with the tests? It would be awesome to have autopilot support :)

**Comment by mirestrepo** [comment]
It would be so nice to see this merging. Seemed like you all were so close! 💪

**Comment by cmcga1125** [comment]
Hey everyone, sorry i've gotten really busy this last month or so and havn't had extra time to look at this.  If anyone wants to pick it up feel free.  I hope to maybe look at it in the next few weeks.

I went through the codelab and it was helpful, just havn't had time to work on the tests.

**Comment by splieth** [comment]
I wonder how the `ip-mask-agent` is supposed to work in autopilot mode. As far as I understood, one is not allowed to alter any resource in the `kube-system` namespace (according to [the docs](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview#managed_namespaces)) and thus creating a config map there for the `ip-mask-agent` is IMHO not possible.

**Comment by philippeboyd** [comment]
Any news on this?

**Comment by morgante** [comment]
@philippeboyd We're still waiting for someone to pick it up from @cmcga1125. Feel free to take it over!

**Comment by odise** [comment]
@morgante Would it be possible to agree on what's left to do on this PR so that it can be merged? I see some minors such as the node pool comment but also some bigger issues like `var.zones` to be handled. What else needs to be changed from your point of view?

**Comment by FearlessHyena** [comment]
Since a lot of work has gone into this PR maybe it would be better to reopen this?

**Comment by morgante** [comment]
> @morgante Would it be possible to agree on what's left to do on this PR so that it can be merged? I see some minors such as the node pool comment but also some bigger issues like `var.zones` to be handled. What else needs to be changed from your point of view?

The last review's comments need to be addressed and someone will also have to resolve merge conflicts. I'd be very happy to take a look if someone is able to pick this up.

**Comment by odise** [comment]
I'm interested to finish the PR and get it merged. However it would be extremely helpful to have a fast catch-up beforehand to get a feeling how much effort it will be. Especially because the complexity of this module is quite high and raising with the autopilot feature.

**Comment by cmcga1125** [comment]
> I'm interested to finish the PR and get it merged. However it would be extremely helpful to have a fast catch-up beforehand to get a feeling how much effort it will be. Especially because the complexity of this module is quite high and raising with the autopilot feature.

i'd be happy to chat about it, but do not have the ability to run the testing sweet

**Comment by jmymy** [comment]
Spinning up the testing env now and will have a look... hopefully have something by tomorrow

**Comment by jmymy** [comment]
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1148

**Comment by jmymy** [comment]
can close. merged in https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/1148

**Pull request closed without merging**